include/lbann/layers/io/input/input_layer.hpp

```diff
@@ -69,22 +69,37 @@ class input_layer : public generic_input_layer {
   }
 
   data_layout get_data_layout() const override { return T_layout; }
+  El::Device get_device_allocation() const override { return Dev; }
 
 };
 
 template<>
-inline void input_layer<partitioned_io_buffer, data_layout::MODEL_PARALLEL>::validate_data_layout() {
+inline void input_layer<partitioned_io_buffer, data_layout::MODEL_PARALLEL, El::Device::CPU>::validate_data_layout() {
   static_assert(true, "input_layer with partitioned_io_buffer does not supports MODEL_PARALLEL data layout");
 }
 
 template<>
-inline void input_layer<partitioned_io_buffer, data_layout::DATA_PARALLEL>::validate_data_layout() {}
+inline void input_layer<partitioned_io_buffer, data_layout::DATA_PARALLEL, El::Device::CPU>::validate_data_layout() {}
 
 template<>
-inline void input_layer<distributed_io_buffer, data_layout::MODEL_PARALLEL>::validate_data_layout() {}
+inline void input_layer<distributed_io_buffer, data_layout::MODEL_PARALLEL, El::Device::CPU>::validate_data_layout() {}
 
 template<>
-inline void input_layer<distributed_io_buffer, data_layout::DATA_PARALLEL>::validate_data_layout() {}
+inline void input_layer<distributed_io_buffer, data_layout::DATA_PARALLEL, El::Device::CPU>::validate_data_layout() {}
+
+template<>
+inline void input_layer<partitioned_io_buffer, data_layout::MODEL_PARALLEL, El::Device::GPU>::validate_data_layout() {
+  static_assert(true, "input_layer with partitioned_io_buffer does not supports MODEL_PARALLEL data layout");
```

Comment:
This doesn't do anything. I'd change the bool to `false`. This change should be made to the <partitioned_io_buffer, MODEL_PARALLEL, CPU> template specification as well.

Reply:
fixed